Converting Avalara Setting page to React

Akarshit opened the pull request:
I am posting the screenshots of old and new for the community to do a design-review.

__Old__ ->

<img width="371" alt="screen shot 2017-11-20 at 9 37 33 pm" src="https://user-images.githubusercontent.com/7762605/33187834-b647b4b0-d0ba-11e7-9a86-82219469330b.png">
----------------------
<img width="376" alt="screen shot 2017-11-20 at 9 37 43 pm" src="https://user-images.githubusercontent.com/7762605/33187833-b61ebc0e-d0ba-11e7-92a1-9dcdd4e3025c.png">
=============================

__New__ ->

<img width="383" alt="screen shot 2017-11-24 at 1 51 23 am" src="https://user-images.githubusercontent.com/7762605/33187825-a60f2f24-d0ba-11e7-9835-e4722d17bc73.png">
--------------------------
<img width="375" alt="screen shot 2017-11-24 at 1 52 02 am" src="https://user-images.githubusercontent.com/7762605/33187824-a5e611d4-d0ba-11e7-8024-d747ee2be6f1.png">

zenweasel commented:
@Akarshit I think if possible we should have this expand into the larger panel size like Email
Also some linting errors there

Akarshit commented:
@zenweasel It would really look weird if Avalara Settings get a larger panel and other tax settings(like TaxCloud) get a smaller panel.
Also don't worry about the error I am still trying to integrate this work in the best way. Will do the linting once done.

Akarshit commented:
The PR is working now and is open for code/design review.
## Fix for #3334 
## To test
1. Login as admin
2. Enable Avalara from the `Tax` right hand menu item.
3. Test the form.
4. But some item and complete the checkout.
5. Test the logging in Avalara(The enable logging button, and clicking on the log should provide it's detail).

zenweasel commented:
@Akarshit One thing I can see w/o even pulling the code is that you need to be able to select multiple values under "Enable Address Validation by Country" when you only have a select box there.

Generally though I think this is a good start. Are you ready for me to look at the code?

zenweasel commented:
Also I don't agree that it would look weird for Avalara to get a larger panel. I actually think it looks more weird to have those settings (and the log viewer) and forced into that tiny space. I would be fine if that means we expand taxcloud to be larger as well.

zenweasel commented:
Change the title to reflect it's WIP status

zenweasel commented:
Okay, I see how you can select multiple values there, but we definitely would want to get some design input because that's a pretty funky UX.

zenweasel commented on imports/plugins/core/ui/client/components/forms/form.js:
```diff
@@ -1,22 +1,41 @@
 import React, { Component } from "react";
 import PropTypes from "prop-types";
-import { map, update, set, at, isEqual } from "lodash";
+import { at, get, isEqual, map, set, update, extend } from "lodash";
```
Is there any need for `map` or `extend` in an ES6 world?

Akarshit replied:
Not really, just that I thought it would look consistent if all object/array functions are used from the same package and have similar syntax while being used.
Though not very attached to this, let me know and I will change it.

zenweasel replied:
Yes, you are correct that we should be consistent and also follow the style guide. So we want to eliminate usages of `lodash` whenever there is an ES6-native equivalent.
https://docs.reactioncommerce.com/reaction-docs/master/styleguide

zenweasel commented:
@Akarshit Well, we are waiting for you. Is this ready for review?

Akarshit commented:
Yes it is, that's why I removed the `WIP` from the title.

zenweasel commented:
Ok, but your first comment says "This is not finished" so you see how I could be unclear? 😄

Akarshit commented:
@zenweasel Oh yes I forgot to update the description. Sorry :)

zenweasel commented:
@rymorgan Where did we leave the design review with this? Should we go ahead and merge it and redesign that component separately?

rymorgan commented:
@zenweasel Yea, I think that's ok, for now. There are probably two places within the multi-product management (which I'm working on now) where I'll be using a multi-select component. I'll design it out there and we can apply that here.

zenweasel commented:
@Akarshit Can you deal with the lint errors and conflicts and I will try to get this approved asap?

Akarshit commented:
@zenweasel I added lines to both the files but Codacy still says a `eol-last` problem. In my local editor there is a empty line at last in both the files.

zenweasel commented:
@Akarshit When I run `eslint .` locally I get no errors and I also show a blank line in those files so it looks like this might be a problem with Codacy.

zenweasel commented:
Did we lose the link to the Admin Console along the way here?

![2017-12-22_09-00-24](https://user-images.githubusercontent.com/72819/34280999-96409a18-e6f6-11e7-9d01-cca87165a2e7.png)

zenweasel commented:
When I enable transaction logging but still have no transactions the logging panel looks a little funky.

zenweasel commented:
Wrong name is in success toast.

![2017-12-22_09-16-01](https://user-images.githubusercontent.com/72819/34281306-d1e4675a-e6f8-11e7-8614-5e506e716b58.png)

Akarshit commented:
@zenweasel 
> Did we lose the link to the Admin Console along the way here?

I changed it to show only when the API credentials are not there. This is consistent with TaxCloud form.
<img width="379" alt="screen shot 2017-12-22 at 2 21 13 pm" src="https://user-images.githubusercontent.com/7762605/34291715-88d4ea00-e723-11e7-9807-8b0bb6e529a5.png">

> Wrong name is in success toast.

Not able to reproduce this. I get the right name when updating the settings.

> When I enable transaction logging but still have no transactions the logging panel looks a little funky.

Fixed

zenweasel commented:
@Akarshit The requirements from Avalara state the the link to the admin console needs to always be there at all times.

Akarshit commented:
Updated the PR with a permanent link to Avalara Console.

zenweasel commented:
Still seeing this toast here my man:

![2018-01-04_10-47-41](https://user-images.githubusercontent.com/72819/34548378-c4bc8352-f13c-11e7-8e86-aa016d8e832b.png)

zenweasel commented:
You have a reference to `admin.update.updateSucceeded` but the only place that exists is in the TaxCloud plugin. Not sure how you would ever see anything else:

https://github.com/reactioncommerce/reaction/blob/1f0e1ff07f476178012d7122b242ac5e83666580/imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js#L39

Akarshit commented:
@zenweasel Updated the PR with i18n fixes.

zenweasel commented:
@Akarshit Please address my code comment regarding `lodash` and we should be ready to go with this

Akarshit commented:
@zenweasel Replaces the lodash functions with native equivalents.

zenweasel commented on .gitignore:
```diff
@@ -12,7 +12,7 @@
 *.sublime-project
 *.sublime-workspace
-
+.package-lock.json
```
Wait, I don't think we want to ignore the package-lock.json anymore

Akarshit commented:
@zenweasel I have removed `package-lock.json` from `.gitignore`

zenweasel commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+    super(props);
+
+    this.state = {
+      editingId: undefined,
```
What's the thinking around settings this value to explicitly be `undefined`? Not saying it's wrong but it seems weird to define it as undefined

Akarshit replied:
Not really, I think this is something that I missed in the cleanup.

aldeed replied:
I do usually define all state properties in the constructor, but I set to `null`, which seems to be the pattern I see most. You don't technically need to do it, but it helps provide a complete list of everything you track in `state` for anyone reading the code.

zenweasel commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+    this.props.handleTestCredentials(event, this.props.settings);
+  }
+
+  editRow = (options) => {
```
JSDoc?

zenweasel commented on imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js:
```diff
@@ -0,0 +1,153 @@
+          return Alerts.toast(i18next.t("settings.testCredentialsSuccess"), "success");
+        }
+        return Alerts.toast(i18next.t("settings.testCredentialsFailed"), "error");
+      } catch (e) {
```
Don't use single letter variables. Always use `error`. If you can't because of scope then use `err`

zenweasel commented on imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js:
```diff
@@ -0,0 +1,153 @@
+    "settings.addressValidation.countryList": {
+      renderComponent: "multiselect",
+      options: validCountries,
+      defaultValue: _.map(validCountries, "value")
```
Why do we use the underscore `_map` here rather than the native?

zenweasel commented on imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js:
```diff
@@ -0,0 +1,153 @@
+    * @param {String} targetField - where to save the new settings in the Avalara Package.
+    * @return {null} - returns nothing
+    */
+  handleSubmit(event, changedInfo, targetField) { // eslint-disable-line no-unused-vars
```
I feel like this UI stuff should be moved out of the container and into the component if possible. Generally I would like containers to contain only data.

zenweasel commented:
@aldeed If you have a few minutes maybe you could take a look at this PR?

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+
+    this.state = {
+      editingId: undefined,
+      showLogs: this.props.settings.avalara.enableLogging,
```
Use `props` instead of `this.props` in constructors

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+        "settings.avalara.enableLogging": {
+          handleChange: this.handleLogToggle
+        }
+      })
```
It shouldn't be necessary to clone `fieldsProp` beyond one level, so I think it's nicer to do it this way:

```js
fieldsProp: {
  ...props.fieldsProp,
  "settings.avalara.enableLogging": {
    handleChange: this.handleLogToggle
  }
}
```

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  handleLogToggle = () => {
+    this.setState(() => ({ showLogs: !this.state.showLogs }));
```
You should pass just the object here, not a function that returns the object.

Akarshit replied:
@aldeed My intention was to write `this.setState((prevState) => ({ showLogs: !prevState.showLogs }));`. Fixing it.

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+        editingId: id,
+        log
+      };
+    });
```
Why passing a function to setState rather than just having those lines before the `setState` call?

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+        {!settings.avalara.apiLoginId &&
+          <div className="alert alert-info">
+            <Components.Translation defaultValue="Add API Login ID to enable" i18nKey="admin.taxSettings.avalaraCredentials" />
+            <a href="https://admin-development.avalara.net" target="_blank"> Avalara</a>
```
Remove the space before "Avalara". Let all spacing be determined by styles.

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+        <Form
+          schema={AvalaraPackageConfig}
+          doc={{ settings }}
+          renderFromFields={true}
```
`={true}` isn't needed for boolean props

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+          doc={{ settings }}
+          renderFromFields={true}
+          fieldsProp={this.state.fieldsProp}
+          docPath={"settings.avalara"}
```
Do not wrap string prop values in `{}`

prinzdezibel replied:
@aldeed Would doing so considered to be bad practice? Just asking, because Webstorm is inserting brackets automatically and it would be tedious to remove them every time I'm going to insert string literals. Not sure if this is configurable.

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+        <div className="panel-body text-center avalara-login-box">
+          <a href="https://admin-development.avalara.net" target="_blank">Avalara Admin Console Login</a>
+        </div>
+        {this.state.showLogs &&
```
In JSX, you need to make sure each condition is exactly `true` or `false` rather than just truthy or falsy. So always use `!!`. (`{!!this.state.showLogs &&`)

aldeed commented on imports/plugins/included/taxes-avalara/client/components/avalaraSettingsForm.js:
```diff
@@ -0,0 +1,182 @@
+          />
+        )}
+        <div>
+          {this.state.log &&
```
!!

aldeed commented on imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js:
```diff
@@ -0,0 +1,153 @@
+    "settings.avalara.companyId"
+  ];
+
+  validCountries = validCountries || Countries.find({ value: { $in: countryDefaults } }).fetch();
```
A `find` in minimongo is pretty much the same as reading the local variable, so I would not bother with caching the result in `validCountries`.

aldeed commented on imports/plugins/included/taxes-avalara/client/containers/avalaraSettingsFormContainer.js:
```diff
@@ -0,0 +1,153 @@
+  const shownLogFields = {
+    date: LogSchema._schema.date,
+    data: LogSchema._schema.data
+  };
```
Everything above here in this function is static. Can you remove it from here and use `withProps` to pass them in instead? Then this composer function would only set `settings` prop whenever it reruns.

Akarshit commented:
@aldeed Thanks for the review.
Will be fixing everything :)

mikemurray commented:
@Akarshit Have you been able to make the updates?

Akarshit commented:
@mikemurray Moving the handler from container to component is a little hard here. Working on that change, except that everything else is done and working.
I will try to complete this ASAP.

spencern commented:
Thanks for the update @Akarshit.

zenweasel commented:
Making this WIP until @Akarshit returns and can resolve issues

Akarshit commented:
@zenweasel Updated the PR by incorporating all the suggestions.

aldeed commented:
@Akarshit Thanks for the changes. We are enforcing additional eslint rules now, so I suggest that you also merge `release-1.8.0` branch into your branch, and then run `npm install` followed by `npm run lint` and fix the lint errors. Some errors can fix themselves if you run `npx eslint . --fix`, so it's good to start with that and then fix the rest manually. Going forward, also make sure your code editor is set up to read the eslint config and show you any errors while you're editing.

aldeed commented on imports/plugins/core/ui/client/components/forms/form.js:
```diff
@@ -243,24 +288,33 @@ class Form extends Component {
+            const fieldProp = get(fieldsProp, key, undefined);
+            return this.renderField({ fieldName: key }, Object.assign(tempObj, fieldProp));
           }
+          return;
```
You do not need to disable `consistent-return` if you change `return` to `return null`.

Also, it is more readable to do early returns rather than large `if` blocks. Something like:

```js
if (!fieldSchema) return null;
// Remove inherited type() as type is supposed to be string.
if (typeof tempObj.type === "function") {
  delete tempObj.type;
}
const fieldProp = get(fieldsProp, key, undefined);
return this.renderField({ fieldName: key }, Object.assign(tempObj, fieldProp));
```

Also, I think it's a bit better to use `{}` as the `fieldProp` default rather than undefined since you are passing it to Object.assign:

```js
const fieldProp = get(fieldsProp, key, {});
```

aldeed commented on imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:
```diff
@@ -27,12 +27,12 @@ taxCalc.getPackageData = function () {
     baseUrl = "https://sandbox-rest.avatax.com/api/v2/";
   } else {
-    baseUrl = "https://rest.avatax.com";
+    baseUrl = "https://rest.avatax.com/api/v2/";
```
It is always better (more understandable and faster code) to do fewer `!`, so you should flip this `if` logic:

```js
if (mode) {
  baseUrl = "https://rest.avatax.com/api/v2/";
} else {
  baseUrl = "https://sandbox-rest.avatax.com/api/v2/";
}
```

Akarshit commented:
@aldeed updated.

zenweasel commented:
These tests are passing for me locally

aldeed commented:
Error on Circle is

```
docker tag reactioncommerce/reaction:5113ea8c1054d436f25d2cb17c551d3f718d9199 reactioncommerce/reaction:pull/3348 
Error parsing reference: "reactioncommerce/reaction:pull/3348" is not a valid repository/tag: invalid reference format
Exited with code 123
```

I'm not sure what would cause that. Maybe temporary issue. I'm ok with merging this.

spencern commented:
@Akarshit Not passing tests. I just rebuilt, but can't merge until that's resolved.

spencern commented:
I'd be okay merging this if I was certain that it wouldn't cause tests to fail on the release branch going forward. As it is, this branch is failing and other branches are passing, pointing to this branch as the issue. Let's merge release-1.8.0 into this branch and see if that resolves the tests

Akarshit commented:
Merged `release-1.8.0` into this.

Akarshit commented:
@spencern This failed at the docker image deployment step
```
#!/bin/bash -eo pipefail
.circleci/deploy.sh
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username: EOF
Exited with code 1
```

spencern commented:
Interesting.. @ticean looks like we'll need to account for forked PRs in the docker push step

spencern commented:
@Akarshit in the mean time, I'm going to merge this as it's passing all tests and the docker build failure is a config issue.
